File: 0668-66-2698-find-the-punishment-number-of-an-integer/0668-66-2698-find-the-punishment-number-of-an-integer.java
class Solution {
    private boolean solve(int num, int target) {
        if (num == target) return true;
        if (num < target || target < 0) return false;
        return solve(num / 10, target - num % 10) ||
               solve(num / 100, target - num % 100) ||
               solve(num / 1000, target - num % 1000);
    }

    public int punishmentNumber(int n) {
        int ans = 0;
        for (int i = 1; i <= n; i++) {
            if (solve(i * i, i)) {
                ans += i * i;
            }
        }
        return ans;
    }

    // Example Usage:
    // Solution sol = new Solution();
    // System.out.println(sol.punishmentNumber(n));
}

File: 0668-66-2698-find-the-punishment-number-of-an-integer/0668-66-2698-find-the-punishment-number-of-an-integer.py
class Solution:
    def punishmentNumber(self, n: int) -> int:
        ans = 0
        def solve(num, target):
            if num == target:
                return True
            if num < target or target < 0:
                return False
            return solve(num//10,target - num%10) or solve(num//100,target - num%100) or solve(num//1000,target - num%1000)
        for i in range(1,n+1):
            if solve(i*i,i):
                ans+=i*i
        return ans